docs/ONNXIFI.md

```diff
@@ -20,13 +20,19 @@ We propose a cross-platform API for loading and executing ONNX graphs on optimiz
 
-0. (Optional) Use `onnxifi_library_load` to dynamically load the ONNX Interface for Framework Integration library.
-1. Call `onnxGetNumBackends` to get the number of available backends. Note that it can be 0.
+0. (Optional) Use `onnxifi_load` to dynamically load the ONNX Interface for Framework Integration library.
+1. Call `onnxGetBackendIDs` to get stable identifiers of available backends. Note: it is possible there are no backends installed in the system.
 2. Call `onnxGetBackendInfo` to check additional information about any available backend.
-3. Call `onnxGetBackendCompatibility` to check if your model, or parts of it, can run on the backend.
+3. Call `onnxGetBackendCompatibility` to check which operations within your model can run on the backend.
 4. Call `onnxInitBackend` to initialize a backend, then call `onnxInitGraph` to offload one or more model graphs to the backend.
-5. Call `onnxSetGraphIO` to set inputs and output for the graph, then call `onnxRunGraph` to execute the graph(s). If your model works with fixed-size inputs, one call to `onnxSetGraphIO` is sufficient for multiple `onnxRunGraph` calls. For models with variable-size inputs, you'd need to call `onnxSetGraphIO` before each `onnxRunGraph` call.
-6. When done using the model, release the model graph(s) with `onnxReleaseGraph`, then release the backend with `onnxReleaseBackend`
+5. Call `onnxSetGraphIO` to set locations and shapes for inputs and outputs of a graph.
+6. Initialize an `inputFence` structure of type `onnxMemoryFenceV1`: set `tag` to `ONNXIFI_TAG_MEMORY_FENCE_V1`, `type` to `ONNXIFI_SYNCHRONIZATION_EVENT`, and call `onnxInitEvent` to initiaze the `event` member.
+7. Initialize an `outputFence` structure of type `onnxMemoryFenceV1`: set `tag` to `ONNXIFI_TAG_MEMORY_FENCE_V1`, `type` to `ONNXIFI_SYNCHRONIZATION_EVENT`, and `event` to null.
+8. Call `onnxRunGraph` with the initialized `inputFence` and `outputFence` structures to enable execution of the graph. The call to `onnxRunGraph` will populate `event` member of the `outputFence` with a newly created event object, asynchronously execute the graph once `inputFence`'s `event` is signalled, and then signal the `outputFence`'s `event`.
+9. Call `onnxSignalEvent` with `event` member of `inputFence` to signal to the backend that the inputs are ready to be consumed. 
+10. Call `onnxWaitEvent` (alternatively, repeatedly call `onnxGetEventState` in a loop until the event state is `ONNXIFI_EVENT_STATE_SIGNALLED`) with `event` member of `outputFence` to wait until graph outputs are ready to be consumed. Release events for inputs and outputs using `onnxReleaseEvent`.
+11. If your model works with fixed-size inputs and outputs, or shape and location of inputs and outputs does not change, one call to `onnxSetGraphIO` is sufficient for multiple `onnxRunGraph` calls. For models with variable-size inputs or outputs, you'd need to call `onnxSetGraphIO` before each `onnxRunGraph` call.
```

Comment:
how can you run `onnxRunGraph` multiple times without `onnxSetGraphIO`? I assume you can only do that **if** previous run is completed. Can you mention that there is no expected concurrent execution of `onnxRunGraph` with the same `setGraphIO`?

Reply:
@rdzhabarov I think onnxSetGraphIO only gives the IO info to backend, therefore it is possible to onnxRunGraph for multiple times but onnxSetGraphIO for only one time? Like
onnxSetGraphIO(...);
//first time running
onnxRunGraph(...);
onnxSignalEvent(...);
onnxWaitEvent(...);
... //get the output, maybe even modified the input descriptor...
//then run a second time without changing the shape and address of input&output
onnxRunGraph(...);
...